magma/circuit.py

```diff
@@ -271,16 +266,20 @@ def __new__(metacls, name, bases, dct):
 
         # Create interface for this circuit class.
         cls._syntax_style_ = _SyntaxStyle.NONE
-        IO = _get_interface_decl(cls)
+        IO = _get_interface_type(cls)
         if IO is not None:
             cls.IO = IO
-            cls.interface = cls.IO(defn=cls,
-                                   renamed_ports=dct["renamed_ports"])
-            _setattrs(cls, cls.interface.ports)
+            cls.interface = cls.IO(defn=cls, renamed_ports=dct["renamed_ports"])
+            setattrs(cls, cls.interface.ports, lambda k, v: isinstance(k, str))
         try:
             context = _definition_context_stack.pop()
             assert context.placer.name == cls_name
-            cls._context_ = context.finalize(cls)
+            try:
+                context = dct["_context_"]
+            except KeyError:
+                pass
+            cls._context_ = context
```

Comment:
could simplify to `cls._context_ = dct.get("_context_", context)`

Reply:
Done.